### ISSUE TITLE
Add support for closures in the frontend

### ISSUE THREAD

**Issue by xuwei-k** [scala-native/scala-native#862]
- https://travis-ci.org/xuwei-k/scalaz/builds/245650378#L1524
- https://github.com/xuwei-k/scalacheck/tree/0a88a0104d5f5f
- https://github.com/xuwei-k/scalaz/commit/2cdc658e160f46fb6b3664c2ff452e52307e53ce

```
scala.scalanative.util.UnsupportedException: (() => anon$1.this.$anonfun$1(fa, f)) (class scala.reflect.internal.Trees$Function)
	at scala.scalanative.util.package$.unsupported(package.scala:17)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genExpr(NirCodeGen.scala:660)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genApplyTypeApply(NirCodeGen.scala:1999)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genApply(NirCodeGen.scala:1081)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genExpr(NirCodeGen.scala:568)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genBlock(NirCodeGen.scala:917)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genExpr(NirCodeGen.scala:621)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$88.apply(NirCodeGen.scala:2135)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$88.apply(NirCodeGen.scala:2135)
	at scala.scalanative.nir.Focus$$anonfun$sequenced$1.apply(Focus.scala:91)
	at scala.scalanative.nir.Focus$$anonfun$sequenced$1.apply(Focus.scala:90)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at scala.scalanative.nir.Focus$.sequenced(Focus.scala:90)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genSimpleArgsWithPt(NirCodeGen.scala:2135)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genMethodArgs(NirCodeGen.scala:2170)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genMethodCall(NirCodeGen.scala:2114)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genMethodCall(NirCodeGen.scala:2096)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genApply(NirCodeGen.scala:1104)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genExpr(NirCodeGen.scala:568)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$19.apply(NirCodeGen.scala:515)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$19.apply(NirCodeGen.scala:515)
	at scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:41)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genNormalMethodBody(NirCodeGen.scala:514)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$genMethod$1.apply(NirCodeGen.scala:304)
	at scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:41)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genMethod(NirCodeGen.scala:279)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$genMethods$1.apply(NirCodeGen.scala:265)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$genMethods$1.apply(NirCodeGen.scala:263)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genMethods(NirCodeGen.scala:263)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genNormalClass(NirCodeGen.scala:193)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$genClass$1.apply(NirCodeGen.scala:166)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$genClass$1.apply(NirCodeGen.scala:164)
	at scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:41)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genClass(NirCodeGen.scala:164)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$apply$4.apply(NirCodeGen.scala:143)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$apply$4.apply(NirCodeGen.scala:139)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.apply(NirCodeGen.scala:139)
	at scala.tools.nsc.Global$GlobalPhase$$anonfun$applyPhase$1.apply$mcV$sp(Global.scala:467)
	at scala.tools.nsc.Global$GlobalPhase.withCurrentUnit(Global.scala:458)
	at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:467)
	at scala.tools.nsc.Global$GlobalPhase$$anonfun$run$1.apply(Global.scala:425)
	at scala.tools.nsc.Global$GlobalPhase$$anonfun$run$1.apply(Global.scala:425)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:425)
	at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.run(NirCodeGen.scala:115)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1528)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1513)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1508)
	at scala.tools.nsc.Global$Run.compile(Global.scala:1609)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:116)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:95)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:107)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:53)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
	at sbt.compiler.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:49)
	at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
	at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:32)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:72)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:71)
	at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:99)
	at sbt.inc.Incremental$.compile(Incremental.scala:71)
	at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
	at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
	at sbt.Compiler$.compile(Compiler.scala:155)
	at sbt.Compiler$.compile(Compiler.scala:141)
	at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:886)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:877)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:875)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by densh** [comment]
Thanks for the bug report! 👍

**Comment by densh** [comment]
Minimization:
```
// build.sbt
scalacOptions ++= Seq("-Ybackend:GenBCode", "-Ydelambdafy:method", "-target:jvm-1.8")

// app.scala
object Test {
  def main(args: Array[String]): Unit =
    Nil.foreach(println)
}
```
With these options Function node survives all the way until backend and is not desugared by the prior phases. We crash because we expect functions to be desugared prior to NIRCodeGen phase.

**Comment by densh** [comment]
@xuwei-k Would it work for you not to omit `-Ydelambdafy:method` for Scala Native builds for now? This is a non-trivial issue to fix, it's on the roadmap for `0.4`, because the same problem is present on 2.12.x due to new LambdaMetafactory-based encoding for closures.

**Comment by WojciechMazur** [comment]
Outdated, minimization provided by @densh works as expected in 2.11.
